### PR TITLE
Masked scale meta function registration

### LINKED ISSUE #119983: Using masked_scale in a torch.compile graph without meta registration breaks the graph
### 🐛 Describe the bug

Consider a test like the one below for a AOT compile backend (such as Habana "hpu").
def test_masked_scale():
	def fn(input, mask, val):
               t1 = torch.add(input, 1)
               t2 = torch.relu(t1)
               t3 = torch._masked_scale(t2, mask, val)
		t4 = torch.sub(t3, 1)
		t5 = torch.relu(t4)
		return t5
 
	input = torch.randn((3, 4))
	mask = torch.randn((3, 4)) > 0.5
	val = 100.0
	h_input = input.to("hpu")
	h_mask = mask.to("hpu")
	# HPU
	compiled_fn = torch.compile(fn, backend="aot_hpu_training_backend")
	hresult = compiled_fn(h_input, h_mask, val)
	
Running this test with TORCH_COMPILE_DEBUG=1 shows a graph break happening due to masked_scale not having a meta registration.

[2024-01-23 08:15:04,373] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] Graph break: unsupported operator: aten._masked_scale.default (see https:/
  /docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit#heading=h.64r4npvq0w0 for how to fix) from user code at:
 [2024-01-23 08:15:04,373] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     t3 = torch._masked_scale(input, mask, val)
 [2024-01-23 08:15:04,373] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]
 [2024-01-23 08:15:04,373] [0/0] torch._dynamo.output_graph: [DEBUG] restore_graphstate: removed 1 nodes
 [2024-01-23 08:15:04,373] [0/0] torch._dynamo.output_graph: [DEBUG] COMPILING GRAPH due to GraphCompileReason(reason='unsupported operator: aten._masked_scale.de
  fault (see
https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit#heading=h.64r4npvq0w0
for how to fix)', user_stack=[<FrameSummary

This becomes a performance issue due to graph break. Adding a meta registration for masked_scale should resolve this issue.

### Versions

python collect_env.py
Collecting environment information...
PyTorch version: 2.1.1a0+git5f7d2b6
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.5 (ssh://gerrit.habana-labs.com:29418/tpc_llvm10 5bbd479866a937ff3ae8772d37b4f1c38e3deb59)
CMake version: version 3.25.2
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 1
Core(s) per socket:                 6
Socket(s):                          2
Stepping:                           0
BogoMIPS:                           4389.68
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  VMware
Virtualization type:                full
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           12 MiB (12 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] habana-torch-dataloader==1.15.0+gitef82bfac3
[pip3] habana-torch-plugin==1.15.0+gitef82bfac3
[pip3] numpy==1.26.3
[pip3] torch==2.1.1a0+git5f7d2b6
[pip3] torch_tb_profiler==0.4.0
[pip3] torchaudio==2.1.0+6ea1133
[pip3] torchdata==0.7.0+c5f2204
[pip3] torchmetrics==1.2.1
[pip3] torchtext==0.16.0a0+4e255c9
[pip3] torchvision==0.16.0+fbb4cc5
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] numpy                     1.21.5           py37h6c91a56_3  
[conda] numpy-base                1.21.5           py37ha15fc14_3  
[conda] pytorch                   1.10.2          cpu_py37hfa7516b_0  
[conda] torchvision               0.11.3          cpu_py37h164cc8f_2

### PR DESCRIPTION
**Pull request by satheeshhab**
Fixes #119983 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @aakhundov

### CODE REVIEW

**Comment by satheeshhab** [comment]
Can this be reviewed please?

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -3796,6 +3796,12 @@ def meta_masked_scatter_backward(self, mask, sizes):
     return self.new_empty(sizes)
 
 
+@register_meta(aten._masked_scale.default)
+def _masked_scale(self, mask, scale):
```
Comment (soulitzer) [inline torch/_meta_registrations.py:3800]:
```suggestion
def meta__masked_scale(self, mask, scale):
```

**Review by soulitzer — COMMENTED** [review]
Thanks, this looks good to me
I added a small comment

**Comment by satheeshhab** [comment]
Updated patch based on review comment from @soulitzer

**Comment by Dalian991** [comment]
-- 
Adrian Rosales

From: Jeffrey Wan ***@***.***> ***@***.***>
Reply: pytorch/pytorch
***@***.***>
***@***.***>
Date: February 19, 2024 at 8:35:19 PM
To: pytorch/pytorch ***@***.***>
***@***.***>
CC: Subscribed ***@***.***>
***@***.***>
Subject:  Re: [pytorch/pytorch] Masked scale meta function registration (PR
#119984)

***@***.**** commented on this pull request.

**Comment by Dalian991** [comment]
-- 
Adrian Rosales

From: satheeshhab ***@***.***> ***@***.***>
Reply: pytorch/pytorch
***@***.***>
***@***.***>
Date: February 20, 2024 at 1:01:56 AM
To: pytorch/pytorch ***@***.***>
***@***.***>
CC: Subscribed ***@***.***>
***@***.***>
Subject:  Re: [pytorch/pytorch] Masked scale meta function registration (PR
#119984)

Updated patch based on review comment from @soulitzer

**Review by soulitzer — APPROVED** [review]
Thanks for the update

**Comment by soulitzer** [comment]
Whoops, looks like it wasn't able to automatically rebase. Do you mind doing the rebase manually and force pushing?

**Comment by soulitzer** [comment]
It means your change, by side-effect, might have fixed something that previously failed, e.g. some logic in quantization depends on the masked scale function is now able to be compiled. The test suite asserts that it is failing to make sure the state is being explicitly tracked. You should update dynamo_test_failures.py to remove the tests that are now passing from the list of tests expected to fail.

**Comment by soulitzer** [comment]
Hmm actually this might be a preexisting, failure. https://github.com/pytorch/pytorch/pull/120271 should fix it.

**Comment by satheeshhab** [comment]
ok, will wait for #120271 to come in

**Review by zou3519 — COMMENTED** [review]
can you add a test please?

**Comment by satheeshhab** [comment]
@zou3519 - We actually encountered this while testing masked_scale with dynamic=True set on Habana "hpu" backend. The test used looks like the following and I am guessing that this can probably be adapted for other dummy test backends too.

def test_masked_scale():
    def fn(input, mask, val):
        t1 = torch.add(input, 1)
        t2 = torch.relu(t1)
        t3 = torch._masked_scale(t2, mask, val)
        t4 = torch.sub(t3, 1)
        t5 = torch.relu(t4)
        return t5

    input = torch.randn((3, 4))
    mask = torch.randn((3, 4)) > 0.5
    val = 100.0
    h_input = input.to("hpu")
    h_mask = mask.to("hpu")
    # HPU
    compiled_fn = torch.compile(fn, backend="aot_hpu_training_backend", dynamic=True)
    hresult = compiled_fn(h_input, h_mask, val)

**Comment by satheeshhab** [comment]
@zou3519 @soulitzer - is there still a problem with this change?

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -3797,6 +3797,12 @@ def meta_masked_scatter_backward(self, mask, sizes):
     return self.new_empty(sizes)
 
 
+@register_meta(aten._masked_scale.default)
+def meta__masked_scale(self, mask, scale):
+    masked_scale = self.new_empty(self.size())
```
Comment (peterbell10) [inline torch/_meta_registrations.py:3802]: This should be created with `memory_format=utils.suggest_memory_format(self)`.

You also need to rebase to fix some of the flaky tests, but many are genuine test failures caused by tests marked as xfail passing.  You need to update the xfails.

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: torch/_meta_registrations.py (hunk 2 above)
Comment (peterbell10) [reply torch/_meta_registrations.py]: You still need to fix the xpasses.
